Document process (#1)
cloudposse/reference-architectures: README.md

# reference-architectures

> Get up and running quickly with one of our reference architecture using our cold start process. 

Reviewer: ```suggestion
Get up and running quickly with one of our reference architectures using our cold start process. 
```

or 
```
Get up and running quickly with our reference architecture using our cold start process.
```

  - [Table of Contents](#table-of-contents)
  - [Known Limitations](#known-limitations)
  - [High Level Overview](#high-level-overview)
  - [Architecture](#architecture)
    - [Assumptions](#assumptions)
    - [Checklist](#checklist)
    - [1. Provision Root Account](#1-provision-root-account)
    - [2. Provision Subaccounts](#2-provision-subaccounts)
    - [3. Delegate DNS](#3-delegate-dns)
  - [Next Steps](#next-steps)
  - [Getting Help](#getting-help)

## Known Limitations

- AWS does not support programmatic deletion of accounts. This means that if you use this project to create the account structure, terraform is not able to completely destroy it.
- AWS by default only permits one subaccount. This limit can be easily increased for your organization, but can take up to several days.

## High Level Overview

You can provision the basic reference architecture in 3 "easy" steps. =)

All accounts will leverage our [`terraform-root-modules`](https://github.com/cloudposse/terraform-root-modules/) to get started.

This process involves using terraform to generate the code (`Dockerfile`, `Makefile`, `terraform.tfvar`, etc) you will use to manage your infrastructure. 

> This is a "bootstrap" process. You do it once and then you throw *this* repo away.

Reviewer: Maybe we need to elaborate why `you throw *this* repo away` (literally or figuratively)

## Architecture

Our "reference architecture" is an opinionated approach to architecting accounts for AWS. 

This process provisions (7) accounts which have different designations. 

Here is what it includes.

| Account | Description                                                                          |
| ------- | ------------------------------------------------------------------------------------ |
| root    | The "root" (parent) account creates all child accounts and is where users login      |
| prod    | The "production" is account where you run your most mission critical applications    |
| staging | The "staging" account is where you run all of your QA/UAT/Testing                    |
| dev     | The "dev" sandbox account is where you let your developers have fun and break things |

> | audit   | The "audit" account is where all logs end up.                                        |

Reviewer: ```suggestion
| audit   | The "audit" account is where all logs end up                                         |
```

| corp    | The "corp" account is where you run platform services for the company                |
| data    | The "data" account is where the quants live =)                                       |

Each account has its own [terraform state backend](https://github.com/cloudposse/terraform-aws-tfstate-backend), along with a [dedicated DNS zone](https://www.terraform.io/docs/providers/aws/r/route53_zone.html) for service discovery.

The root account owns the top-level DNS zone and then delegates NS authority to each child account.

### Assumptions

1. We are starting with a clean AWS environment and a new "root" (top-level) AWS account. This means you need the "root" credentials, since a fresh AWS account doesn't even have any AWS roles that can be assumed.
2. You have administrator access to this account
3. You have [docker](https://docs.cloudposse.com/tools/docker/) installed on your workstation
4. You have [terraform](https://www.terraform.io/downloads.html) installed on your workstation

### Checklist

Before we get started, make sure you have the following

> - [ ] Before you can create new AWS accounts under your organization, you must [verify your email address](https://docs.aws.amazon.com/console/organizations/email-verification).

Reviewer: ```suggestion
- [ ] Before you can create new AWS accounts under your organization, you must [verify your email address](https://docs.aws.amazon.com/console/organizations/email-verification)
```

> - [ ] Open a support ticket to [request the limit](https://console.aws.amazon.com/support/v1#/case/create) of AWS accounts be increased for your organization (the default is 1).

Reviewer: ```suggestion
- [ ] Open a support ticket to [request the limit](https://console.aws.amazon.com/support/v1#/case/create) of AWS accounts be increased for your organization (the default is 1)
```

- [ ] Clone this repo on your workstation

> - [ ] Create a *temporary* pair of [Access Keys](https://console.aws.amazon.com/iam/home#/security_credential). These should be deleted afterwards.

Reviewer: ```suggestion
- [ ] Create a *temporary* pair of [Access Keys](https://console.aws.amazon.com/iam/home#/security_credential). These should be deleted afterwards
```

- [ ] Export your AWS "root" account credentials as `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` (this is *temporary* for bootstrapping)

> - [ ] An available domain we can use for DNS-base service discovery (E.g. `ourcompany.co`). This domain must not be in use else where.

Reviewer: ```suggestion
- [ ] An available domain we can use for DNS-base service discovery (E.g. `ourcompany.co`). This domain must not be in use elsewhere
```

> - [ ] Ensure that any users who will be added during this bootstrap process have setup their [keybase](http://keybase.io) profile. You'll need this if setting them up in the `users` section of the [`config/root.tfvars`](https://github.com/cloudposse/reference-architectures/blob/master/configs/root.tfvars). For example you should be able to [verify their public key](https://keybase.io/osterman/key.asc) on `keybase.io` by running `curl https://keybase.io/$username/key.asc`.

Reviewer: ```suggestion
- [ ] Ensure that any users who will be added during this bootstrap process have setup their [keybase](http://keybase.io) profile. You'll need this if setting them up in the `users` section of the [`config/root.tfvars`](https://github.com/cloudposse/reference-architectures/blob/master/configs/root.tfvars). For example you should be able to [verify their public key](https://keybase.io/osterman/key.asc) on `keybase.io` by running `curl https://keybase.io/$username/key.asc`
```

### 1. Provision Root Account

The "root" account is the top-most AWS account from which all other AWS accounts are programmatically created.

> __WARNING:__ Terraform cannot remove an AWS account from an organization. Terraform will not close the account. The member account must be prepared to be a standalone account beforehand. See the [AWS Organizations documentation](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_accounts_remove.html) for more information.

Reviewer: Above it says `all other AWS accounts are programmatically created`, but then `The member account must be prepared to be a standalone account beforehand`. More info needed on why and how?

> __WARNING:__ Do not chain the `make` targets together (e.g. `make root children finalize` as it is not currently supported)

Reviewer: ```suggestion
__WARNING:__ Do not chain the `make` targets together (e.g. `make root children finalize`) as it is not currently supported
```

5. Create the IAM groups to permit access to child accounts
6. Write a list of child account IDs so we can use them in the next phase

</details>

### 2. Provision Subaccounts

Subaccounts are created by the root account, but are ultimately provisioned using the subaccount containers.

Update the configuration for all the child accounts by editing the `configs/root.tfvar` file.

To get started, run:

```bash
make children
```

<details>

<summary>Here's what that roughly looks like (but entirely automated).</summary>

For each child account:

1. Create a new account git repo
2. Render the templates for a `child` account into the repo directory (include `Dockerfile`). Obtain the account ID from the previous phase.
3. Build a docker image
4. Run the docker image and start provisioning the child account's Terraform state bucket, DNS zone, cloudtrail logs, etc.

</details>

### 3. Delegate DNS

Now that each subaccount has been provisioned, we can delegate each DNS zone to those accounts.

To finish up, run:

```bash
make finalize
```

<details>
<summary>Here's what that roughly looks like (but entirely automated).</summary>

> 1. Re-use the docker image from phase (1) and phase (2)

Reviewer: ```suggestion
1. Re-use the docker images from phase (1) and phase (2)
```

2. Update DNS so that root account delegates DNS zones to the child accounts
3. Enable cloudtrail log forwarding to audit account

</details>

---

## Next Steps

At this point, you have everything you need to start terraforming your way to success.

All of your account configurations are currently in `repos/`

1. Commit the changes in `repos/`. Open Pull Requests.

2. Ensure that the name servers for the service discovery domain (e.g. `ourcompany.co`) have been configured with your domain registrar (e.g. GoDaddy)

> 3. Delete your root account credentials. They are no longer need and should not be used. Instead use IAM users.

Reviewer: ```suggestion
3. Delete your root account credentials. They are no longer needed and should not be used. Instead use the created IAM users.
```
